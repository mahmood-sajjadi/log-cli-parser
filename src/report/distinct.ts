export function distinctInStream<I> (list: I[], value: I): I[] {
  if (list.some(item => item === value)){
    return list;
  } else {
    // can be modified to address immutability with cost of performance
    list.push(value);
    return list;
  }
}

/**
 * @deprecated Use `distinctInStream` instead. This function will be removed in version 2.0.
 */
export function distinct<R, I> (arr: R[], fn: (item: R) => I): I[] {
  return arr.map(item => fn(item)).filter((value: I, index: number, array: I[]) => {
    return array.indexOf(value) === index;
  })
}
