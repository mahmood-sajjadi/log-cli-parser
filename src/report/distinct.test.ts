import { distinctInStream, distinct } from './distinct';

describe('distinctInStream', () => {
  test('WHEN item EXISTS in list do not ADD', () => {
    const list = ['abc', 'mno'];
    const result = distinctInStream(list, 'abc');
    expect(result).toHaveLength(2);
    expect(result).toMatchSnapshot();
  });

  test('WHEN item NOT EXISTS in list ADD', () => {
    const list = ['abc', 'mno'];
    const result = distinctInStream(list, 'xyz');
    expect(result).toHaveLength(3);
    expect(result).toMatchSnapshot();
  });
});

describe('distinct', () => {
  test('WHEN list PROVIDED then CREATE distinct result', () => {
    const list = [{a: 'abc', b: 10}, {a: 'mno', b: 100}, {a: 'abc', b: 3}];
    const result = distinct(list, (item) => item.a);
    expect(result).toHaveLength(2);
    expect(result).toMatchSnapshot();
  });
});
