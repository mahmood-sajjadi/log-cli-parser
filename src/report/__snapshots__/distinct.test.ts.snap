// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`distinct WHEN list PROVIDED then CREATE distinct result 1`] = `
[
  "abc",
  "mno",
]
`;

exports[`distinctInStream WHEN item EXISTS in list do not ADD 1`] = `
[
  "abc",
  "mno",
]
`;

exports[`distinctInStream WHEN item NOT EXISTS in list ADD 1`] = `
[
  "abc",
  "mno",
  "xyz",
]
`;
